lib/puppet/functions/each.rb

```diff
@@ -1,38 +1,76 @@
 #
-# When the first argument is an Array (or of enumerable type other than Hash), the parameterized block
-# should define one or two block parameters.
-# For each application of the block, the next element from the array is selected, and it is passed to
-# the block if the block has one parameter. If the block has two parameters, the first is the elements
-# index, and the second the value. The index starts from 0.
+# @example Using the `each` function
 #
-#       $a.each |$index, $value| { ... }
-#       each($a) |$index, $value| { ... }
+# `$each_data = $data.each |$parameter| { <PUPPET CODE BLOCK> }`
 #
-# When the first argument is a Hash, the parameterized block should define one or two parameters.
-# When one parameter is defined, the iteration is performed with each entry as an array of `[key, value]`,
-# and when two parameters are defined the iteration is performed with key and value.
+# or
 #
-#       $a.each |$entry|       { ..."key ${$entry[0]}, value ${$entry[1]}" }
-#       $a.each |$key, $value| { ..."key ${key}, value ${value}" }
+# `$each_data = each($data) |$parameter| { <PUPPET CODE BLOCK> }`
 #
-# @example using each
+# When the first argument (`$data` in the above example) is an array, Puppet passes each
+# value in turn to the lambda, then returns the original values.
 #
-#       [1,2,3].each |$val| { ... }                       # 1, 2, 3
-#       [5,6,7].each |$index, $val| { ... }               # (0, 5), (1, 6), (2, 7)
-#       {a=>1, b=>2, c=>3}].each |$val| { ... }           # ['a', 1], ['b', 2], ['c', 3]
-#       {a=>1, b=>2, c=>3}.each |$key, $val| { ... }      # ('a', 1), ('b', 2), ('c', 3)
-#       Integer[ 10, 20 ].each |$index, $value| { ... }   # (0, 10), (1, 11) ...
-#       "hello".each |$char| { ... }                      # 'h', 'e', 'l', 'l', 'o'
-#       3.each |$number| { ... }                          # 0, 1, 2
+# @example Using the `each` function with an array and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the array $data, run a lambda outputting each value multiplied by 10 using the
+# # notice function, then return the array
+# $data = [1, 2, 3]
+# $each_data = $data.each |$item| { notice ($item * 10) }
+# # The notice function runs three times, outputting 10, then 20, then 30
+# # $each_data contains [1, 2, 3]
+# ~~~
+#
+# When the first argument is a hash, Puppet passes each key and value pair to the lambda
+# as an array in the form `[key, value]` and returns a hash containing the results.
+#
+# @example Using the `filter` function with a hash and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the hash $data, run a lambda using each item as a key-value array and outputting
+# # the keys using the notice function, then return the hash
+# $data = {"a" => 0, "b" => 1, "c" => 2}
+# $each_data = $data.each |$items| { notice ($items[0]) }
+# # The notice function runs three times, outputting a, then b, then c
+# # $each_data contains {a => 0, b => 1, c => 2}
+# ~~~
```

Comment:
Here's an interesting exercise for you: can you think of an example that uses a hash to construct resources? Maybe in the same family of tasks (using data to create symlinks)?